### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by janisozaur**
I added basic test set up, since I don't really know how to best wire it all up, I need your input.

I used google test, packaged for Arch. It is current release, 1.8.0, Ubuntu has 1.6 in the repos, I'm not sure what would we be missing, but [1.6 is quite dated](https://github.com/google/googletest/commit/800f5422ac9d9e0ad59cd860a2ef3a679588acb4) (18 Apr 2011) and it may cause conflicts with C++11. I haven't checked.

This is set up as a sibling directory to `test/testpaint`. Things which would need resolving:
* Are you guys happy with google test?
* Should there be one binary per test or should they all be linked into one? Former may be quicker to build, better support parallel testing and easier to hook up when working with specific parts of code, latter may be easier to maintain.
  * I think I could make both configurations possible based on configure-time switch.
* Should we try to support Ubuntu and its package?
  * It may be more fruitful to backport a more recent package for one of the jobs.
* Port `testpaint` to google test?
* Write better `CMakeLists.txt` for our tests, include it in the main one.
* Write more tests.

Right now it's the most basic check, not complete just yet. It tests that a roundtrip through sawyercoding functions results in the input message. I'll add more scenarios to the test, right now this is proof of concept submitted for discussion.

But hey, it works:

```
Running main() from gtest_main.cc
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from sawyercoding
[ RUN      ] sawyercoding.write_read_chunk_none
[       OK ] sawyercoding.write_read_chunk_none (0 ms)
[ RUN      ] sawyercoding.write_read_chunk_rle
[       OK ] sawyercoding.write_read_chunk_rle (1 ms)
[ RUN      ] sawyercoding.write_read_chunk_rle_compressed
[       OK ] sawyercoding.write_read_chunk_rle_compressed (1 ms)
[ RUN      ] sawyercoding.write_read_chunk_rotate
[       OK ] sawyercoding.write_read_chunk_rotate (0 ms)
[----------] 4 tests from sawyercoding (2 ms total)
[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (2 ms total)
[  PASSED  ] 4 tests.
```

### CODE REVIEW

File: test/tests/sawyercoding_test.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+extern "C" {
+	#include "util/sawyercoding.h"
+}
+
+#include <gtest/gtest.h>
+
+// TODO: Seed with some proper randomness
```
Comment (marijnvdwerf) [inline test/tests/sawyercoding_test.cpp]: Shouldn't tests be fixed?

File: test/tests/sawyercoding_test.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+char randomdata[] = { 0x01, 0x02, 0x03, 0x04 };
```
Comment (marijnvdwerf) [inline test/tests/sawyercoding_test.cpp]: Why does the file use `char`, `uint8`, and `uint8_t`? Shouldn't one of those be preffered?

**Comment by marijnvdwerf** [comment]
I wasn't satisfied with googletest's parameterized tests for testpaint. The output was pretty uncomfortable naming-wise. So googletest for testpaint gets a 'meh'/'why bother' from me.

**Comment by marijnvdwerf** [comment]
I remember reading somewhere that you shouldn't depend on googletest in repositories, and that the recommended way of using it is by including it. You might want to check https://github.com/OpenRCT2/OpenRCT2/pull/4330/commits/efc306b2afc3b7746be3cb5dca31b937bed0b8a9, which did work iirc.

**Comment by janisozaur** [comment]
Yes, I had it configured with `ExternalProject_Add` a while back, but Arch updated their package since. Do you remember where have you read that? Caching it on Travis would be nice if we decide to go this way.

**Comment by marijnvdwerf** [comment]
I updated my comment with a link

**Comment by IntelOrca** [comment]
How about using a class to represent the sawyer coding test fixture instead with each test being a method of the class?

**Comment by janisozaur** [comment]
@marijnvdwerf  OK, I'm partially swayed. Partially, because there's `gtest-config`, which you can use the flags gtest was compiled with.

**Comment by marijnvdwerf** [comment]
```raw
brew search googletest
Installing gtest system-wide is not recommended; it should be vendored
in your projects that use it.
```

**Comment by janisozaur** [comment]
I've updated `CMakeLists.txt` to use `ExternalProject_add`, gtest is now downloaded & compiled on demand.

How do you like it now?

**Comment by janisozaur** [comment]
I've made some more improvements to the tests setup. They are now hooked up to main `CMakeLists.txt` file, guarded by configuration-time switch `WITH_TESTS` (default: `off`). There's one job that compiles and runs the tests.

Let me know what you think.

**Comment by IntelOrca** [comment]
In sawyercoding_test.cpp, can you move the random data arrays to the bottom of the file or to another file so that one isn't required to scroll past it to see the code?

**Comment by janisozaur** [comment]
Yes. I'll also change the predefined data to less random one, probably extract something from some custom object.

**Comment by janisozaur** [comment]
I moved the data definition to the end of file, but left it intact otherwise, I'm not concerned with changing it after all.

I think it should be ready to go, at least as far as whole set up goes.

**Comment by marijnvdwerf** [comment]
Shouldn't it match the guidelines of C++ code indenting and file-naming wise?

**Comment by janisozaur** [comment]
Reformatted

File: test/tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		LOG_BUILD ON)
+
+
+# Specify include dir
+set(GTEST_INCLUDE_DIR ${GOOGLETEST_DISTRIB_SOURCE_DIR}/googletest/include)
+
+# Library
+ExternalProject_Get_Property(googletest BINARY_DIR)
+set(GOOGLETEST_BINARY_DIR "${BINARY_DIR}")
+set(GTEST_LIBRARY_PATH ${GOOGLETEST_BINARY_DIR}/${CMAKE_FIND_LIBRARY_PREFIXES}gtest.a)
```
Comment (marijnvdwerf) [inline test/tests/CMakeLists.txt]: `${CMAKE_STATIC_LIBRARY_PREFIX}gtest${CMAKE_STATIC_LIBRARY_SUFFIX}`?

File: test/tests/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+set(GTEST_MAIN_LIBRARY_PATH ${GOOGLETEST_BINARY_DIR}/${CMAKE_FIND_LIBRARY_PREFIXES}gtest_main.a)
+set(GTEST_LIBRARY gtest)
+set(GTEST_MAIN_LIBRARY gtest_main)
+add_library(${GTEST_LIBRARY} UNKNOWN IMPORTED)
```
Comment (marijnvdwerf) [inline test/tests/CMakeLists.txt]: `STATIC` instead of `UNKNOWN`?

File: test/tests/CMakeLists.txt (hunk 3 above)
Comment (janisozaur) [inline test/tests/CMakeLists.txt]: I can even make it `${CMAKE_STATIC_LIBRARY_PREFIX}${GTEST_LIBRARY}${CMAKE_STATIC_LIBRARY_SUFFIX}` :P

File: test/tests/LanguagePackTest.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    ASSERT_EQ(lang->GetId(), 1);
+    ASSERT_EQ(lang->GetCount(), 1);
+    ASSERT_STREQ(lang->GetString(0), "");
+    lang->SetString(0, "xx");
+    ASSERT_EQ(lang->GetCount(), 1);
+    ASSERT_STREQ(lang->GetString(0), "xx");
+    delete lang;
+}
+
+extern const utf8 * language_en_GB;
```
Comment (IntelOrca) [inline test/tests/LanguagePackTest.cpp]: `gLanguageEnGB`

Comment (marijnvdwerf) [inline test/tests/LanguagePackTest.cpp]: or `LanguageBasic`? As opposed to multibyte languages.

**Pull request merged**